condor_git_config.py

```diff
@@ -91,127 +98,174 @@ class ConfigCache(object):
         self.git_uri = git_uri
         self.branch = branch
         self.cache_path = cache_path
         self.max_age = max_age
-        self._work_path = os.path.abspath(os.path.join(cache_path, branch))
-        self._meta_file = self.abspath('cache.json')
-        self._cache_lock = filelock.FileLock(
-            os.path.join('/tmp', '.cache.%s.lock' % zlib.adler32(os.path.basename(__file__).encode()))
-        )
-        os.makedirs(self._work_path, exist_ok=True, mode=0o755)
+        self._work_path = cache_path.resolve() / branch
+        self._work_path.mkdir(mode=0o755, parents=True, exist_ok=True)
+        self._meta_file = self.abspath("cache.json")
+        self._cache_lock = filelock.FileLock(str(self.abspath(f"cache.{branch}.lock")))
+        self._config_meta: Optional[Dict[str, Any]] = None
 
-    def abspath(self, *rel_paths):
-        return os.path.abspath(os.path.join(self._work_path, *rel_paths))
+    @property
+    def stats(self):
+        """Statistics of the cache"""
+        meta = self._config_meta if self._config_meta is not None else self._read_meta()
+        age = time.time() - meta["timestamp"]
+        return {"age": age, "pulls": meta["pulls"], "reads": meta["reads"]}
 
-    def repo_path(self, *rel_paths):
-        return self.abspath('repo', *rel_paths)
+    def _read_meta(self):
+        try:
+            with open(self._meta_file, "r") as meta_stream:
+                meta_data = json.load(meta_stream)
+        except FileNotFoundError:
+            return {
+                "git_uri": self.git_uri,
+                "branch": self.branch,
+                "timestamp": 0.0,
+                "pulls": 0,
+                "reads": 0,
+            }
+        else:
+            if meta_data["git_uri"] != self.git_uri:
+                LOGGER.critical("cache %r corrupted by other hook: %r", self, meta_data)
+                raise RuntimeError(
+                    "config cache %r used for conflicting hooks" % self.cache_path
+                )
+            return meta_data
+
+    def _write_meta(self):
+        if self._config_meta is None:
+            return
+        with open(self._meta_file, "w") as meta_stream:
+            json.dump(self._config_meta, meta_stream)
+
+    def abspath(self, *rel_paths: Union[str, Path]) -> Path:
+        return self._work_path.joinpath(*rel_paths)
+
+    def repo_path(self, *rel_paths: Union[str, Path]) -> Path:
+        return self.abspath("repo", *rel_paths)
 
     def __enter__(self):
         self._cache_lock.acquire()
+        self._config_meta = self._read_meta()
+        self._refresh()
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
+        self._write_meta()
+        self._config_meta = None
         self._cache_lock.release()
         return False
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable[Path]:
+        assert self._config_meta is not None
```

Comment:
You might also consider dropping this assertion as you should always enter in a valid state with a `_config_meta` that is not `None`.

Reply:
These assertions tell the type checker that the attribute is indeed set in these methods. (FWIW, I agree that they aren't useful as defensive checks against a human messing up.)